src/java/org/apache/cassandra/db/filter/RowFilter.java

```diff
@@ -230,6 +246,13 @@ public String toString()
         return root.toString();
     }
 
+    private void assertFilterIsNotATree()
```

Comment:
RowFilter used to be a list of Expressions. It is still being used as a list in many places using `getExpressions` or `iterator()`.
However, RowFilter may be a tree now, and ^ methods were not updated to reflect that fact.
Thus, to be on the safe side, we assert that any use of RowFilter in list context must be carried out only on flat RowFilters.

Ideally, we'd change the uses of RowFilter etc., but that is out of scope

Reply:
@mike-tr-adamson can you review the changes here?

Reply:
As I've said below, this was, originally, tested in such a way that it was functionally correct according to the tests in place. If this change breaks any tests then it has broken the functionality.

I know that's a bit cryptic but the whole `StatementRestrictions` / `RowFilter` is a mess in this regard and only really works because the tests say it works.

Reply:
So.. I'm seeing the assertion fire in:
```
java.lang.IllegalStateException: RowFilter must be a flat list of Expressions to be used in any iterative context like iterator() or getExpressions()
	at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
	at org.apache.cassandra.db.filter.RowFilter.assertFilterIsNotATree(RowFilter.java:251)
	at org.apache.cassandra.db.filter.RowFilter.iterator(RowFilter.java:239)
	at org.apache.cassandra.index.SecondaryIndexManager.getBestIndexQueryPlanFor(SecondaryIndexManager.java:1172)
	at org.apache.cassandra.db.ReadCommand.findIndexQueryPlan(ReadCommand.java:356)
```
and
```
java.lang.IllegalStateException: RowFilter must be a flat list of Expressions to be used in any iterative context like iterator() or getExpressions()
	at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
	at org.apache.cassandra.db.filter.RowFilter.assertFilterIsNotATree(RowFilter.java:251)
	at org.apache.cassandra.db.filter.RowFilter.hasExpressionOnClusteringOrRegularColumns(RowFilter.java:110)
	at org.apache.cassandra.db.SinglePartitionReadCommand.selectsFullPartition(SinglePartitionReadCommand.java:1079)
```

To me, this means that finding a query plan may not work correctly.
@mike-tr-adamson, can you confirm that in these cases it is assertion that is a problem, not the executed code path?
If so, I'll change the assertion to no spam log warning...

Reply:
I managed to cause assertion failure with a test.
In order not to delay this PR I'm changing the assertion to a log and also restoring the original behaviour of the filter.
The issue with the assertion failure etc will be reported separately